### PR TITLE
Reap deployments and ReplicationControllers

### LINKED ISSUE #2316: Deployer pod isn't deleted when the deployment is
When deleting a replication contoller in the cli, its respective deployer pod stays around.

cc: @ironcladlou @smarterclayton

### PR DESCRIPTION
**Pull request by ironcladlou**
Implement reaping for deployments and ReplicationControllers, reusing
upstream reapers wherever possible.

Closes #2316

### CODE REVIEW

**Comment by ironcladlou** [comment]
cc @abhgupta @kargakis @smarterclayton

**Comment by ironcladlou** [comment]
[test]

**Comment by smarterclayton** [comment]
Why do we have to specifically reap controllers?

**Comment by smarterclayton** [comment]
Won't the deployment pod controller handle the delete of the rc?  Why do we have to customize the reaper

**Comment by ironcladlou** [comment]
>Why do we have to specifically reap controllers?

We already delegate ReplicationController reaping to the upstream implementation (although indirectly via the DeploymentConfig reaper). I maintain that here (but organized differently).

File: pkg/cmd/util/clientcmd/factory.go [hunk 1]
```diff
@@ -321,3 +330,51 @@ func (c *clientCache) ClientForVersion(version string) (*client.Client, error) {
...
+	}
+
+	// Set up an upstream controller reaper.
+	controllerReaper, err := kubectl.ReaperForReplicationController(r.kc, r.timeout)
+	if err != nil {
+		return "", err
+	}
+
+	var reaper kubectl.Reaper
+	if deployutil.IsDeployment(controller) {
```
Comment (smarterclayton) [inline pkg/cmd/util/clientcmd/factory.go]: Specifically why do you need to do this.

**Comment by ironcladlou** [comment]
>Won't the deployment pod controller handle the delete of the rc? Why do we have to customize the reaper

We preserve failed deployment RCs and their deployer pods for inspection. Only completed deployments and their deployers are automatically deleted.

**Comment by ironcladlou** [comment]
I feel like we're debating the validity of #2316 rather than the implementation details of this PR.

**Comment by smarterclayton** [comment]
The deployment delete pod controller is responsible for deleting deployer pods and hook pods.  The reaper deleting the rc is enough.  If there are extra scenarios describe them.

**Comment by ironcladlou** [comment]
>The deployment delete pod controller is responsible for deleting deployer pods and hook pods. The reaper deleting the rc is enough. If there are extra scenarios describe them.

The DeploymentController deletes all deployer pods for a completed deployment. The DeployerPodController scales back failed deployments and scales up the prior completed deployment on deployment failure.

No controllers will delete deployer pods for a failed or deleted deployment. Failed deployer pods aren't deleted because the user wants to inspect the failed stuff, and deleted deployments don't trigger any controller-based because the controllers aren't aware of deletions in general except in one special case (if a deployer pod disappears for a running deployment, the deployment is moved to failed).

The only way to get rid of the failed deployer pods in the current architecture is to delete them directly, or delete them during the cascading delete of the deployment to which the deployer pods are associated. Hence the existence of #2316. What am I missing here?

**Comment by smarterclayton** [comment]
> On Jun 12, 2015, at 10:19 AM, Dan Mace <notifications@github.com> wrote:
> 
> The deployment delete pod controller is responsible for deleting deployer pods and hook pods. The reaper deleting the rc is enough. If there are extra scenarios describe them.
> 
> The DeploymentController deletes all deployer pods for a completed deployment. The DeployerPodController scales back failed deployments and scales up the prior completed deployment on deployment failure.
> 
> No controllers will delete deployer pods for a failed or deleted deployment.
> 
Failed yes.  Deleted, that doesn't make sense.

**Comment by ironcladlou** [comment]
>> No controllers will delete deployer pods for a failed or deleted deployment.

>Failed yes.  Deleted, that doesn't make sense.

Is this what you're suggesting: We should extend the special cased delete handling in the DeployerPodController such that when the deployment for a deployer pod disappears, the deployer is deleted.

**Comment by smarterclayton** [comment]
> On Jun 12, 2015, at 10:26 AM, Dan Mace <notifications@github.com> wrote:
> 
> No controllers will delete deployer pods for a failed or deleted deployment.
> 
> Failed yes. Deleted, that doesn't make sense.
> 
> Is this what you're suggesting: We should extend the special cased delete handling in the DeployerPodController such that when the deployment for a deployer pod disappears, the deployer is deleted.
> 
Yes.

**Comment by ironcladlou** [comment]
I think it'll work. Good idea. Too bad I didn't think of it before I started this PR.

**Comment by ironcladlou** [comment]
Sorry, PR. You had a good run. 
:skull:

**Comment by ironcladlou** [comment]
I was wrong about resusing the existing poller logic. The way it works is to look at all cached deployments in a Store and check for vanished deployers. What we need for this cleanup use case is to poll all deployer pods in a namespace and validate their deployments still exist.

I don't think that logic fits in with any of the existing controller flow. It would just be a totally separate polling goroutine started by a controller factory which would wake up and do the validation/deletion. Is that okay? It's definitely a departure from the way things work now but I don't see another way yet given we don't use a Delta based mechanism in the controllers.

**Comment by ironcladlou** [comment]
cc @abhgupta @smarterclayton ^

**Comment by ironcladlou** [comment]
Some options:

1. Keep a store populated with deployer pods, validate them periodically
2. Validate in the main DeployerPodController Handle method, since that will fire on the reflector resync period as well and eventually allow us to search for the missing deployment and self-destruct. This has a flaw though in that if the deployer pod is deleted for some reason, and then the deployment is deleted, hook pods could get orphaned since the RC reaper won't get them and neither the DeployerPodController nor the DeploymentController has anything to react to.

**Comment by ironcladlou** [comment]
Speaking of stores, how have we managed to justify using them in our controllers? Aren't they massive memory leaks barring some disk (or other) spooling capability?

**Comment by abhgupta** [comment]
> Option 2. Validate in the main DeployerPodController Handle method, ...

I favor this approach - the deployments are cached locally so, a "get" on the deployment for each deployer pod will not be an issue. Eventually, we would like our controllers to use delta queue and be able handle deletion events.

**Pull request closed without merging**